File: web/frontend/src/api/creatures.ts
import { get, put, post, httpDelete } from './client'

export interface DialogueChoice {
  text: string
  id: string
  next: DialogueLine[]
}

export interface DialogueLine {
  text: string
  speaker: string
  choices?: DialogueChoice[]
}

export interface DialogueEntry {
  lines: DialogueLine[]
}

export interface SoundEntry {
  type: string // attack, defend, greet
  path: string
}

export interface RosterEntry {
  creature_id: string
  level: number
}

export interface Creature {
  name: string
  description: string
  types: string[]
  base_hp: number
  base_attack: number
  base_defense: number
  base_sp_attack: number
  base_sp_defense: number
  base_speed: number
  base_exp: number
  class: string
  category: string
  evolution?: { creature_id: string; level: number; flavor: string }
  recruit_method?: string
  recruit_chance?: number
  recruit_dialogue?: string
  has_overworld_sprite?: boolean
  has_battle_sprite?: boolean
  learnset: { level: number; move_id: string }[]
  sounds: SoundEntry[]
  npc_sprite?: string
  dialogues?: Record<string, DialogueEntry>
  is_hostile?: boolean
  lead_creature?: RosterEntry | null
  roster?: RosterEntry[]
  reserves?: RosterEntry[]
}

/** Convention-based sprite paths derived from creature ID */
export function spritePath(creatureId: string, variant: 'overworld' | 'battle' = 'overworld'): string {
  const file = variant === 'battle' ? `${creatureId}_battle.png` : `${creatureId}.png`
  return `assets/sprites/creatures/${file}`
}

export const creaturesApi = {
  list: () => get<Record<string, Creature>>('/creatures/'),
  getOne: (id: string) => get<Creature>(`/creatures/${id}`),
  update: (id: string, data: Creature) => put<Creature>(`/creatures/${id}`, data),
  create: (category?: string) => post<{ status: string; creature_id: string }>(`/creatures/${category ? `?category=${category}` : ''}`),
  delete: (id: string) => httpDelete<{ status: string }>(`/creatures/${id}`),
  getViewOrder: () => get<string[]>('/creatures/view-order'),
  selectViewOrder: (id: string) => put<string[]>(`/creatures/view-order/${id}`, {}),
}

File: web/backend/routers/creatures.py
from fastapi import APIRouter, HTTPException

from backend.config import settings
from backend.services.data_service import DataService
from backend.services.git_service import GitService

router = APIRouter(prefix="/api/creatures", tags=["creatures"])
data_svc = DataService()
git_svc = GitService()


@router.get("/")
def list_creatures():
    creatures = data_svc.get_all_creatures()
    sprites_dir = settings.repo_path / "assets" / "sprites" / "creatures"
    for cid, cdata in creatures.items():
        # For NPCs, check npc_sprite path; for creatures, use convention-based paths
        npc_sprite = cdata.get("npc_sprite", "")
        if npc_sprite:
            sprite_path = settings.repo_path / npc_sprite.removeprefix("res://")
            cdata["has_overworld_sprite"] = sprite_path.exists()
        else:
            cdata["has_overworld_sprite"] = (sprites_dir / f"{cid}.png").exists()
        cdata["has_battle_sprite"] = (sprites_dir / f"{cid}_battle.png").exists()
    return creatures


@router.get("/{creature_id}")
def get_creature(creature_id: str):
    creature = data_svc.get_creature(creature_id)
    if not creature:
        raise HTTPException(404, f"Creature '{creature_id}' not found")
    return creature


@router.put("/{creature_id}")
def update_creature(creature_id: str, body: dict):
    if not data_svc.get_creature(creature_id):
        raise HTTPException(404, f"Creature '{creature_id}' not found")
    data_svc.update_creature(creature_id, body)
    return {"status": "updated", "creature_id": creature_id}


@router.post("/")
def create_creature(category: str = "wild"):
    all_creatures = data_svc.get_all_creatures()
    prefix = "new_npc" if category == "npc" else "new_creature"
    n = 1
    while f"{prefix}_{n}" in all_creatures:
        n += 1
    creature_id = f"{prefix}_{n}"

    creature_data = {
        "name": "New NPC" if category == "npc" else "New Creature",
        "description": "",
        "types": ["normal"],
        "base_hp": 50,
        "base_attack": 50,
        "base_defense": 50,
        "base_sp_attack": 50,
        "base_sp_defense": 50,
        "base_speed": 50,
        "base_exp": 50,
        "class": "npc" if category == "npc" else "monster",
        "category": category,
        "learnset": [],
    }
    data_svc.create_creature(creature_id, creature_data)
    return {"status": "created", "creature_id": creature_id}


@router.delete("/{creature_id}")
def delete_creature(creature_id: str):
    if not data_svc.delete_creature(creature_id):
        raise HTTPException(404, f"Creature '{creature_id}' not found")
    return {"status": "deleted", "creature_id": creature_id}
